packages/fast-components-react-base/src/listbox/listbox.tsx

```diff
@@ -0,0 +1,284 @@
+    /**
+     * Return an array of all focusabled elements that are children
+     * of the context menu
+     */
+    private domChildren(): Element[] {
+        return canUseDOM() && this.rootElement.current instanceof HTMLElement
+            ? Array.from(this.rootElement.current.children)
+            : [];
+    }
+
+    /**
+     * Ensure we always validate our internal state on item focus events, otherwise
+     * the component can get out of sync from click events
+     */
+    private handleMenuItemFocus = (e: React.FocusEvent<HTMLElement>): void => {
+        const target: Element = e.currentTarget;
+        const focusIndex: number = this.domChildren().indexOf(target);
+
+        if (this.isDisabledElement(target)) {
+            target.blur();
+            return;
+        }
+
+        if (focusIndex !== this.state.focusIndex && focusIndex !== -1) {
+            this.setFocus(focusIndex, focusIndex > this.state.focusIndex ? 1 : -1);
+        }
+    };
+
+    /**
+     * Sets focus to the nearest focusable element to the supplied focusIndex.
+     * The adjustment controls how the function searches for other focusable elements
+     * if the element at the focusIndex is not focusable. A positive number will search
+     * towards the end of the children array, whereas a negative number will search towards
+     * the beginning of the children array.
+     */
+    private setFocus(focusIndex: number, adjustment: number): void {
+        const children: Element[] = this.domChildren();
+
+        while (inRange(focusIndex, children.length)) {
+            const child: Element = children[focusIndex];
+
+            if (this.isFocusableElement(child)) {
+                child.focus();
+
+                this.setState({
+                    focusIndex,
+                    focussedItemId: child.id === undefined ? "" : child.id,
+                });
+
+                break;
+            }
+
+            focusIndex += adjustment;
+        }
+    }
+
+    /**
+     * Handle the keydown event of the root menu
+     */
+    private handleMenuKeyDown = (e: React.KeyboardEvent<HTMLDivElement>): void => {
+        switch (e.keyCode) {
+            case KeyCodes.arrowDown:
+            case KeyCodes.arrowRight:
+                e.preventDefault();
+                this.setFocus(this.state.focusIndex + 1, 1);
+
+                break;
+
+            case KeyCodes.arrowUp:
+            case KeyCodes.arrowLeft:
+                e.preventDefault();
+                this.setFocus(this.state.focusIndex - 1, -1);
+
+                break;
+
+            case KeyCodes.end:
+                e.preventDefault();
+                this.setFocus(this.domChildren().length - 1, -1);
+
+                break;
+
+            case KeyCodes.home:
+                e.preventDefault();
+                this.setFocus(0, 1);
+
+                break;
+
+            default:
+                if (!e.ctrlKey) {
+                    this.processTypeAhead(e);
+                }
```

Comment:
With multi-select there are quite a few additional expectations for keyboard functionality which are different than the expected interaction model for single select: https://www.w3.org/TR/wai-aria-practices-1.1/#listbox_kbd_interaction

I don't think we need to manage the alternative selection model, but we should support the recommend model.

Reply:
Selection isn't managed by this component, but rather the parent component (likely a select, but we should leave the door open to others).  I think type ahead essentially works the same in both cases.

Reply:
Listbox is valid without a parent - how do I implement a listbox and manage selection? I think we need to document this. It seems like we've built this primarily to support our scenario of select

Reply:
Listbox has been refactored to manage selection and supports the recommended kbd interaction you pointed to above.